book/src/engine/major_concepts.md

```diff
@@ -0,0 +1,119 @@
+# Major concepts
+
+This section goes over a few different concepts that are crucial to
+understanding how `chalk-engine` works, without going over the exact solving
+logic.
+
+## `Context`, `ContextOps`, and `InferenceTable`
+
+### `Context`
+
+The [`Context`] trait is the primary bridge between Chalk internal logic and
+external types. In addition actually *defining* the types (via associated
+types), it also contains associated functions to convert or extract
+information from those types. Overall, this allows the types to be basically
+opaque to the engine internals. Functions in the trait are agnostic to specific
+program or environment details, since they lack a `&self` argument.
+
+To give an example, there is an associated `Goal` type. However, Chalk doesn't
+know how to solve this. Instead, it has to be converted an [`HhGoal`] via the
+`Context::into_hh_goal` function. This will be coverted more in the `Goals`
+section.
+
+### `ContextOps`
+
+The [`ContextOps`] trait contains functions that may specifically require
+information a specific program or environment. For example, the
+`program_clauses` function gives potential ways to prove a `Goal`, but obviously
+ it requires knowing the program (for example, what types, traits, and impls
+ there are). Functions in this trait all take a `&self` argument.
+
+### `InferenceTable`
+
+The [`InferenceTable`] is a super trait to the `UnificationOps`, `TruncateOps`,
+and `ResolventOps`. Each of these contains functions that track the state of
+specific parts of the program. Importantly, these operations can dynamically
+change the state of the logic itself.
+
+## Goals
+
+A "goal" in Chalk can be thought of as "something we want to prove". The engine
+itself understands [`HhGoal`]s. `HHGoal`s consist of the most basic logic,
+such as introducing Binders (`Forall` or `Exists`) or combining goals (`All`).
+On the other hand, `Context::Goal` represents an opaque goal generated
+externally. As such, it may contain any extra information or may be interned.
+When solving a logic predicate, Chalk will lazily convert `Context::Goal`s
+into `HHGoal`s.
```

Comment:
I intend to add some chapters to the type section describing the sorts of goals that appear in our logic etc in a more "semantic way", so maybe we should link over there eventually I guess.